### ISSUE TITLE
mount: crashes with panic: send on closed channel

### ISSUE THREAD

**Issue by agncr** [rclone/rclone#4511]
#### What is the problem you are having with rclone?
Rclone mount just "stops" now and then. I see "panic: sendon closed channel" in my logs and i can't reach the mounted folder any longer.

#### What is your rclone version (output from `rclone version`)
rclone v1.52.2
- os/arch: linux/amd64
- go version: go1.14.4

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)
Ubuntu 20.04.1 LTS

####  Which cloud storage system are you using? (eg Google Drive)
Google Drive

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)
(via systemd)
```
/usr/local/bin/rclone mount \
        --allow-other \
        --dir-cache-time 1000h \
        --drive-use-trash \
        --log-file /var/log/rclone.log \
        --log-level DEBUG \
        --no-modtime \
        --poll-interval 15s \
        --rc \
        --rc-addr :5573 \
        --drive-use-trash \
        --uid=1000 \
        --gid=1000 \
        --stats=0 \
        crypt:/ /mnt/rclone
```

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)
```
... a lot of User Rate Limit Exceeded ...
2020/08/14 13:00:41 DEBUG : pacer: low level retry 6/10 (error googleapi: Error 403: User Rate Limit Exceeded. Rate of requests for user exceed configured project quota. You may consider re-evaluating expected per-user traffic to the API and adjust project quota limits accordingly. You may monitor aggregate quota usage and adjust limits in the API Console: https://console.developers.google.com/apis/api/drive.googleapis.com/quotas?project=xyz, userRateLimitExceeded)
2020/08/14 13:00:41 DEBUG : pacer: Rate limited, increasing sleep to 1.434041097s
2020/08/14 13:00:43 DEBUG : pacer: Reducing sleep to 0s
2020/08/14 13:00:47 DEBUG : Google drive root '': Checking for changes on remote
2020/08/14 13:01:02 DEBUG : Google drive root '': Checking for changes on remote
2020/08/14 13:01:17 DEBUG : Google drive root '': Checking for changes on remote
2020/08/14 13:01:32 DEBUG : Google drive root '': Checking for changes on remote
2020/08/14 13:01:47 DEBUG : Google drive root '': Checking for changes on remote
2020/08/14 13:02:02 DEBUG : Google drive root '': Checking for changes on remote
2020/08/14 13:02:02 DEBUG : crypt: Loaded invalid token from config file - ignoring
2020/08/14 13:02:02 DEBUG : Keeping previous permissions for config file: -rw-rw-rw-
2020/08/14 13:02:02 DEBUG : crypt: Saved new token in config file
2020/08/14 13:02:09 NOTICE: Google drive root '': Disabling ListR to work around bug in drive as multi listing (50) returned no entries
panic: send on closed channel

goroutine 238757 [running]:
github.com/rclone/rclone/backend/drive.(*Fs).listRRunner.func2(0xc009d0c700, 0x32, 0x32, 0xc004b6a420, 0xc00010c700)
        github.com/rclone/rclone/backend/drive/drive.go:1759 +0xa3
created by github.com/rclone/rclone/backend/drive.(*Fs).listRRunner
        github.com/rclone/rclone/backend/drive/drive.go:1757 +0xb85
```

**Comment by robertbaker** [comment]
You exceeded the rate limit, you won't be able to mount until 24hrs has passed.

**Comment by darthShadow** [comment]
Nope, you are confusing the upload/download 24-hour bans with regular API rate-limiting. The messages are slightly different for both:

* Upload/Download 24-hour Ban: Error 403: User rate limit exceeded
* Regular API rate-limiting: Error 403: User Rate Limit Exceeded

Anyway, either of those issues should not result in the mount crashing.

**Comment by ncw** [comment]
It looks like the token expired which caused the recursive directory lister to blow up somehow.

@agncr can you make this happen reliably?

**Comment by agncr** [comment]
Unfortunately no: it just happens.
I'll check my logs (syslog, internet connectivity) when it happens next time, maybe i can find something in there.

I also have a vm on Hetzner, where i'm using mount. It also happened there. I assume it's not really related to my laptop.
There i'm using Debian 10 (debug was not activated).
```
2020/08/15 06:34:36 NOTICE: Google drive root 'backups': Disabling ListR to work around bug in drive as multi listing (50) returned no entries                                                
panic: send on closed channel

goroutine 55042 [running]:
github.com/rclone/rclone/backend/drive.(*Fs).listRRunner.func2(0xc01e466000, 0x32, 0x32, 0xc012f92240, 0xc001b31a40)                                                                          
        github.com/rclone/rclone/backend/drive/drive.go:1759 +0xa3
created by github.com/rclone/rclone/backend/drive.(*Fs).listRRunner
        github.com/rclone/rclone/backend/drive/drive.go:1757 +0xb85
```
Someone any suggestion, how i can fetch more relevant information to debug it?

**Comment by ncw** [comment]
That traceback is probably enough info to fix the problem. It is knowing it is fixed that is difficult if it is hard to reproduce.

**Comment by ncw** [comment]
Here is the code which caused the problem

https://github.com/rclone/rclone/blob/d8144a7e84dc3fb4975adf1bce707a5be672fada/backend/drive/drive.go#L1759

https://github.com/rclone/rclone/blob/d8144a7e84dc3fb4975adf1bce707a5be672fada/backend/drive/drive.go#L1757-L1762

What looked like happend is that `in` got closed before the go routine had finished.

This was very likely caused by cbf3d4356135814921382dd3285d859d15d0aa77 when that code was introduced.

The

>  Disabling ListR to work around bug in drive as multi listing (50) returned no entries

in the logs is a big hint here.

**Comment by ncw** [comment]
Ok I think I've worked out what has happened here...

- rclone engaged a workaround to work around a google drive bug
- one of the listing routines returned an error (which never got to the log)
- this caused the `in` channel to be closed to stop the readers
- however the workaround was recycling stuff into the `in` channel at the time
- hence the panic on closed channel

I've fixed this problem - I'd appreciate it if you'd have a go with it @agncr 

https://beta.rclone.org/branch/v1.52.3-316-g8248df62-fix-4511-drive-panic-beta/ (uploaded in 15-30 mins)

**Comment by agncr** [comment]
I'm running it now for 7 days and none of my mounts crashed on 3 different machines. Looks good.

**Comment by ncw** [comment]
Thanks for testing 😃

I'll reopen this to remind me to merge the fix when I'm at my desk!

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.53
